### ISSUE TITLE
Remote-WSL: node always running with high CPU usage

### ISSUE THREAD

**Issue by zhouyifan0904** [microsoft/vscode-remote-release#3171]
I type "code ." inside the Ubuntu WSL2 terminal to launch VS Code with the Remote-WSL extension. After I do that, no matter what directory I'm in, the "node" process is always present in Windows task manager taking up a constant >10% of CPU and kills my battery life. This much CPU utilization is not expected when I'm working natively from the Windows environment even if I have a React dev server running, let alone when I'm not working on anything related to nodejs.

I'm getting around this issue by launching VS Code from the Windows Explorer by typing "explorer.exe ." from WSL. But of course that way I can only edit files but can't access the Linux environment from VS Code.

**Comment by Mshivam2409** [comment]
I am facing the same bug. The node process of vscode server seems to undergoing memory leakage or something. I no longer can access vscode from wsl.

**Comment by zhouyifan0904** [comment]
> I am facing the same bug. The node process of vscode server seems to undergoing memory leakage or something. I no longer can access vscode from wsl.

There are three versions of Ubuntu in the app store: Ubuntu, Ubuntu 20.04 and Ubuntu 18.04. I switched to using the 20.04 app and that one doesn't have this problem. I don't recommend 18.04 because it never seems to release memory taken by "vmmem" process, but 20.04 is working fine for me so far. Hope it helps!

**Comment by Mshivam2409** [comment]
> > I am facing the same bug. The node process of vscode server seems to undergoing memory leakage or something. I no longer can access vscode from wsl.
> 
> There are three versions of Ubuntu in the app store: Ubuntu, Ubuntu 20.04 and Ubuntu 18.04. I switched to using the 20.04 app and that one doesn't have this problem. I don't recommend 18.04 because it never seems to release memory taken by "vmmem" process, but 20.04 is working fine for me so far. Hope it helps!

Vmmem is for WSL2 only. I am using WSL1 though the version 18.04. I'll try updating to 20.04 and see if it changes anything. I tried manually reinstalling the vscode server but It didn't fix anything. Surprisingly my other Laptop running wsl2 18.04 doesn't seem to have this issue.

**Comment by escape0707** [comment]
See this issue if it helps https://github.com/microsoft/vscode/issues/97568#issuecomment-630822277 .

**Comment by jazzfog** [comment]
Seeing a similar issue, after WSL project open I my vscode-server princess consumes 100% of one of the cores all the time.

**Comment by JorgenSmith** [comment]
@jazzfog I'm getting the same thing on my new Ryzen system. Fresh install of latest Windows 10 Pro only a few days ago. Build is: 
`Ubuntu 20.04.1 LTS (GNU/Linux 4.4.0-19041-Microsoft x86_64)`

I restored my WSL2 image from an exported backup, and have done `sudo apt-get update | sudo apt-get upgrade`.

If I have SMP/HT disabled (12 logical cores), each vscode instance goes 100% on a core. With SMP enabled (24 logical cores) it sits on 1 physical core but bounces between the two logical ones it offers like this:
![image](https://user-images.githubusercontent.com/348953/92421221-d2bc7a80-f1ba-11ea-92b5-7d4582fe9394.png)

I quickly dove into @escape0707 's link but I could mainly see WSL1 comments and I need to learn more about this incompatibility with `glibc` and how it relates to our distribution. However: To quickly confirm if affected, try running `htop` - for me, it won't run, resulting in a blank terminal window.

In my (Ryzen) instance, it adds 17c to CCD1 temp compared to CCD2 - not ideal long term, even with good cooling. 

Will update here if I find out more, but wanted to provide this info as a +1.

**Comment by jazzfog** [comment]
I've discovered that on my system it seems like it is connected to 
```
"typescript.tsserver.experimental.enableProjectDiagnostics": true,
```
property that I've enabled.

It allows to see all the TS issues in the project as opposed to just current file.

Coming from Intellij-based IDE I really got used to see all the errors at all times and IDE without this feature is useless for me.
VSCode has it now but seems like optimization is not there - I can't use it with such a high CPU consumption...

**Comment by JorgenSmith** [comment]
Thanks @jazzfog . In my vscode that setting is not enabled, just FYI. Will let you know if I find anything else.

**Comment by JorgenSmith** [comment]
@jazzfog I've been able to resolve it with the following workaround; I can now launch VSCode from wsl2/terminal with no CPU hogging. This involves rolling back `libc` to a specific older version. Hope this helps.

https://github.com/microsoft/vscode-remote-release/issues/2921#issuecomment-624564199

**Comment by JorgenSmith** [comment]
@jazzfog @aeschli I got package compatibility issues when doing `sudo apt-get update && sudo apt-get upgrade`, and decided to start afresh. After typing `explorer.exe .` to open my `/home/jsmith/` directory in Windows Explorer (I found perf issues when working on the windows file system previously) I copied my require files out including `.ssh` and performed a reset of Ubuntu20 in App settings. Then did the update/upgrade. Afterwards, having restored the backed up data, everything worked as it should - I have both `htop` running, and also CPU usage doesn't spin out of control when using code.

There's a possibility my Ubuntu image started off as a v18 one then got upgraded; I can't quite recall. Just putting it there incase it might be related. Thanks.

**Comment by krodelabestiole** [comment]
same here.

according to htop the processes that use the most resource have to do with TypeScript :
`~/.vscode-server/bin/***/node ~/.vscode-server/bin/***/extensions/node_modules/typescript/lib/tsserver.js`

I have never used TypeScript or any extension specific to it. Still I tried to disable every extensions in VS Code, and to move my default WSL machine from Ubuntu 20.04 to Debian Stretch. With no luck : I still have the same issue.

I think the issue has to do with **node watchers**. CPU usage ramps up as soon as I launch any watcher such as a `gatsby develop` command (plus some out of memory errors only when VS Code is running).

I don't think I ever went into the same issue with other watchers (such as the one Hugo uses).

I think VS Code - remote - WSL is running a node watcher and WSL2 doesn't like having 2 node watchers at the same time at all...

**Comment by Hellysonrp** [comment]
I have a similar issue.
Using WSL2 with debian buster.

![image](https://user-images.githubusercontent.com/5438762/110669857-8b379b00-81ab-11eb-880d-d1a3c9599cc5.png)

Comparing with the windows task manager:

![image](https://user-images.githubusercontent.com/5438762/110670306-026d2f00-81ac-11eb-9f63-130c80be2e87.png)

The drop in the graphs is because I closed vscode to compare the average CPU usage.
When vscode is open, CPU usage in all cores is 60%~90%.

It doesn't seem to be extension related, because it isn't the `extensionHost` that is using CPU.

I'm using Windows 10 Pro 1909 build 18363.1379

**Comment by aeschli** [comment]
@Hellysonrp Thanks for reporting. There was a recent change to use polling for watching files when you open a folder in `/mnt/c`. I suspect that;s the problem. I'd like to track this in a separate issue and created #4649.

**Comment by krsche** [comment]
@aeschli Thanks for pointing me in the right direction!
I'm using Ubuntu-20.04 via WSL-1 on a .code-workspace under symlinked ~/Documents to /mnt/c/...
The issue for me was that I have the following set to prevent the folder renaming issue:
```
    "remote.WSL.fileWatcher.polling": true,
    "remote.WSL.fileWatcher.pollingInterval": 2500,
```

This works fine for smaller workspaces but not mine which has ~170k files in it. This is mostly due to several node_modules folders... Would love if `remote.WSL.fileWatcher.pollingExcludeDirectories` would exist 😄 
I'd even implement it myself if someone points me in the right direction.

**Comment by aeschli** [comment]
@krsche Good idea. We already have a setting to exclude folder from being watched: Check out `files.watcherExclude`. You can also set it as a workspace setting

**Comment by krsche** [comment]
@aeschli Great! How didn't I find this 😅 The default settings actually already exclude `**/node_modules/*/**` there and
I tried to repeat my test but wasn't able to reproduce the high CPU usage this time. 
I'll play around with that setting once I face issues again.

**Comment by aeschli** [comment]
To find out which events are reported:
- Run `Developer: Set Log Level...` and set it `Trace`
- Open `Log (Remote Server)` in the Output view 
- make changes in some folders and see if they are reported or not

**Comment by aplotor** [comment]
im getting this even with `"remote.WSL.fileWatcher.polling": false` in wsl1 ubuntu 22.04.4

![image](https://github.com/user-attachments/assets/83766d36-f75b-4cc9-9870-5a87480ed530)

<br/>

it doesnt happen all the time, but when it does i have to kill vsc from task manager and reload all my vsc windows :c

**Comment by aeschli** [comment]
@aplotor If you go to the 'Detail' page you can get more information about the node processes that are running hot. I'd be interested to know the command line of these node processes.

**Comment by aplotor** [comment]
@aeschli it's blank

![image](https://github.com/user-attachments/assets/78c74bcc-6767-4755-9b15-9e70624d5c90)

<br/>

not sure if helps, but the file location is `C:\Users\{user}\AppData\Local\Packages\CanonicalGroupLimited.Ubuntu_79rhkp1fndgsc\LocalState\rootfs\home\{user}\.vscode-server\bin\38c31bc77e0dd6ae88a4e9cc93428cc27a56ba40\node`

**Comment by alexqhj** [comment]
Any updates on this? This is killing my development. tsserver using 90% cpu and intellisense spends 10 seconds "loading" for every change
